[FLINK-27251][checkpoint] Timeout aligned to unaligned checkpoint barrier in the output buffers

1996fanrui opened the pull request:
## What is the purpose of the change

Support upstream Task switching from Aligned Checkpoint to Unaligned Checkpoint to improve the UC when the back pressure is severe. 
When the barrier cannot be sent from the output buffer to the downstream task within the [execution.checkpointing.aligned-checkpoint-timeout](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/deployment/config/#execution-checkpointing-aligned-checkpoint-timeout), the upstream task switches to UC and takes a snapshot of the data before the barrier in the output buffer.

## Brief change log

- *When broadcasting Barrier, if checkpointOptions.isTimeoutable(), register timer.*
  - *The time when the timer is triggered is: CP triggerTime + alignmentTimeout*
  - *Behavior when the timer is triggered: snapshot all buffers that have not been sent to the downstream before the Barrier*
- *If the UC is enabled and the Aligned Barrier is received, the CP cannot succeed immediately and needs to wait for the outputBufferFuture:*
  - *When all barriers are sent downstream quickly, and an empty output buffer is returned, the CP ends.(when the back pressure isn't severe)*
  - *Or wait for the timer to trigger, snapshot these output buffers before the barrier, and the CP ends.(when the back pressure is severe)*
  - *If the CP is aborted, flink will execute the`channelStateFuture.completeExceptionally(cause)`*


## Verifying this change

This change added tests and can be verified as follows:

  - Finish unit test later.
  - This PR will affect the Unaligned checkpoint benchmark. We can view the benefit after merge [here](http://codespeed.dak8s.net:8000/timeline/#/?exe=1,6&ben=checkpointSingleInput.UNALIGNED_1&env=2&revs=200&equid=off&quarts=on&extr=on).


## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? docs, it need to update some doc of UC.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequest.java:
```diff
@@ -103,18 +104,21 @@ static ChannelStateWriteRequest buildFutureWriteRequest(
+                        CloseableIterator<Buffer> iterator =
+                                CloseableIterator.fromList(buffers, Buffer::recycleBuffer);
+                        iterator.close();
+                    } catch (ExecutionException ignored) {
```
Why do we need those changes? Also why are they in the commit that's adding unit tests?

1996fanrui replied:
### Why do we need those changes?

The writer.fail(e) needs be called after catch ExecutionException. It has been called in action, so discardAction shouldn't do anything. So I ignore the ExecutionException.

###  Also why are they in the commit that's adding unit tests?

I found some unit tests failed during develop unit test due to catch Throwable in action. It will drop some exceptions from `writer.writeOutput`. So I change the Throwable to ExecutionException in action, and improved the catch in discardAction.

pnowojski replied:
You shouldn't be modifying the code that you have added in the same pr in another commit. As in our [coding style guideline](https://flink.apache.org/contributing/code-style-and-quality-pull-requests.html#separate-refactoring-cleanup-and-independent-changes):
> There should be no cleanup commits that fix issues that have been introduced in previous commits of the same PR. Commits should be clean in themselves.


Why do you even need to handle exceptions here? The pre-existing `buildWriteRequest()` does nothing like that.

1996fanrui replied:
Hi @pnowojski .

## Let me describe the conclusion first: we must add catch here.

ChannelStateFuture.completeExceptionally(e) will be executed when abort checkpoint, meanwhile, dataFuture.get() will throw ExecutionException.

If there is no catch here, the action throws an ExecutionException, causing the ChannelStateWriteThread to end. ChannelStateWriteThread is shared, and once it ends, subsequent CPs cannot continue.
Related code:
- `ChannelStateWriteRequestExecutorImpl#run`
- `ChannelStateWriteRequestExecutorImplTest#testRecordsException`

_**So channelStateFuture.completeExceptionally should only cause this CP to fail and should not affect the job. If remove this catch, some ITCases will fail too.**_

## Solution
Actually, ChannelStateCheckpointWriter#runWithChecks has similar logic to avoid the ChannelStateWriteThread to end. When writing data throws Exception, execute writer.fail(e), if not IOException, continue to throw.

ChannelStateCheckpointWriter#runWithChecks code:
```java
private void runWithChecks(RunnableWithException r) {
    try {
        checkState(!result.isDone(), "result is already completed", result);
        r.run();
    } catch (Exception e) {
        fail(e);
        if (!findThrowable(e, IOException.class).isPresent()) {
            rethrow(e);
        }
    }
}
```

So, here are two options:

Option A: catch Exception, if not ExecutionException, continue to throw.

```java
writer -> {
    try {
        for (Buffer buffer : dataFuture.get()) {
            checkBufferIsBuffer(buffer);
            bufferConsumer.accept(writer, buffer);
        }
    } catch (Exception e) {
        writer.fail(e);
        if (!findThrowable(e, ExecutionException.class).isPresent()) {
            rethrow(e);
        }
    }
},
throwable -> {
    try {
        CloseableIterator.fromList(dataFuture.get(), Buffer::recycleBuffer).close();
    } catch (Exception e) {
        if (!findThrowable(e, ExecutionException.class).isPresent()) {
            rethrow(e);
        }
    }
},
```

Option B: just catch ExecutionException

```java
writer -> {
    try {
        for (Buffer buffer : dataFuture.get()) {
            checkBufferIsBuffer(buffer);
            bufferConsumer.accept(writer, buffer);
        }
    } catch (ExecutionException e) {
        writer.fail(e);
    }
},
throwable -> {
    try {
        CloseableIterator.fromList(dataFuture.get(), Buffer::recycleBuffer).close();
    } catch (ExecutionException ignored) {
    }
},
```

The difference between the two is that OptionA calls fail(e) when it catches an exception other than ExecutionException . But I think these exceptions already call fail(e) in ChannelStateCheckpointWriter#runWithChecks , so it doesn't need to be considered here, so I choose OptionB in this PR. 

## Improved Solution

Or OptionB should also have some improvements, if `bufferConsumer.accept(writer, buffer);` throws ExecutionException because `ChannelStateCheckpointWriter#runWithChecks` throws ExecutionException, writer.fail(e) is called twice. So here try catch should only contain dataFuture.get().

OptionC:
```java
writer -> {
    List<Buffer> buffers;
    try {
        buffers = dataFuture.get();
    } catch (ExecutionException e) {
        writer.fail(e);
        return;
    }
    for (Buffer buffer : buffers) {
        checkBufferIsBuffer(buffer);
        bufferConsumer.accept(writer, buffer);
    }
},
throwable -> {
    try {
        CloseableIterator.fromList(dataFuture.get(), Buffer::recycleBuffer).close();
    } catch (ExecutionException ignored) {
    }
},
```

What do you think?

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartitionTest.java:
```diff
@@ -347,6 +357,129 @@ public void testProducerFailedException() {
+        // test without data buffer
+        long checkpointId = 5L;
+        subpartition.add(getTimeoutableBarrierBuffer(checkpointId));
+        assertEquals(1, subpartition.getNumberOfQueuedBuffers());
+        assertEquals(1, subpartition.getChannelStateFutures().size());
+        CompletableFuture<List<Buffer>> channelStateFuture =
+                subpartition.getChannelStateFutures().get(checkpointId);
+        assertFalse(channelStateFuture.isDone());
+
+        ResultSubpartition.BufferAndBacklog barrierBuffer = subpartition.pollBuffer();
+        assertNotNull(barrierBuffer);
+        assertEquals(
+                Buffer.DataType.TIMEOUTABLE_ALIGNED_CHECKPOINT_BARRIER,
+                barrierBuffer.buffer().getDataType());
+
+        assertEquals(0, subpartition.getChannelStateFutures().size());
+        assertEquals(0, subpartition.getNumberOfQueuedBuffers());
+        assertNull(channelStateFuture.get());
+
+        // test with data buffer
+        subpartition.resumeConsumption();
+        checkpointId = 6;
+        subpartition.add(createFilledFinishedBufferConsumer(4096));
+        subpartition.add(getTimeoutableBarrierBuffer(checkpointId));
+        assertEquals(1, subpartition.getChannelStateFutures().size());
+        assertEquals(2, subpartition.getNumberOfQueuedBuffers());
+        channelStateFuture = subpartition.getChannelStateFutures().get(checkpointId);
+
+        assertNotNull(subpartition.pollBuffer());
+        assertFalse(channelStateFuture.isDone());
+        barrierBuffer = subpartition.pollBuffer();
+        assertNotNull(barrierBuffer);
+        assertEquals(
+                Buffer.DataType.TIMEOUTABLE_ALIGNED_CHECKPOINT_BARRIER,
+                barrierBuffer.buffer().getDataType());
+
+        assertEquals(0, subpartition.getChannelStateFutures().size());
+        assertEquals(0, subpartition.getNumberOfQueuedBuffers());
+        assertNull(channelStateFuture.get());
+    }
```
Maybe write a helper method `testWithNDataBuffers(int n, ...)`; and then call it twice:
```
testWithNDataBuffers(0, ...);
testWithNDataBuffers(1, ...);
```
to deduplicate the code?

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartitionTest.java:
```diff
@@ -347,6 +357,129 @@ public void testProducerFailedException() {
+        // test without data buffer
+        long checkpointId = 7L;
+        subpartition.add(getTimeoutableBarrierBuffer(checkpointId));
+        assertEquals(1, subpartition.getChannelStateFutures().size());
+        assertEquals(1, subpartition.getNumberOfQueuedBuffers());
+        CompletableFuture<List<Buffer>> channelStateFuture =
+                subpartition.getChannelStateFutures().get(checkpointId);
+        assertFalse(channelStateFuture.isDone());
```
I see those three assertions repeated over and over again. Maybe extract them to a helper method `assertSubpartitionChannelStateFuturesAndQueuedBuffers(subpartition, 1, 1, false);`?

pnowojski reviewed (commented):
Thanks for the PR @1996fanrui ! I think the code looks good, I've left a couple of comments, mostly nits.

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequest.java:
```diff
@@ -72,6 +74,51 @@ static ChannelStateWriteRequest write(
+                    List<Buffer> buffers = dataFuture.get();
+                    if (buffers == null || buffers.isEmpty()) {
+                        return;
+                    }
```
Same as the previous one

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java:
```diff
@@ -191,9 +205,7 @@ private int add(BufferConsumer bufferConsumer, int partialRecordLength, boolean
-        if (prioritySequenceNumber != -1) {
-            notifyPriorityEvent(prioritySequenceNumber);
-        }
+        notifyPriorityEvent(prioritySequenceNumber);
```
Thanks for your reminder, I will do better in the future contribution.

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java:
```diff
@@ -240,9 +255,171 @@ private boolean processPriorityBuffer(BufferConsumer bufferConsumer, int partial
+                }
+            } catch (IOException e) {
+                channelStateFuture.completeExceptionally(
+                        new IllegalStateException(
```
I wrapped some messages and e to the IllegalStateException. Do you mean that `completeExceptionally(e)` directly?

pnowojski replied:
> Do you mean that completeExceptionally(e) directly?

yes, unless there is a good reason for wrapping into `IllegalStateException` here?

1996fanrui commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -103,6 +112,14 @@ class SubtaskCheckpointCoordinatorImpl implements SubtaskCheckpointCoordinator {
+
+    /** Hold the AlignmentTimer for each checkpointId. */
+    @GuardedBy("lock")
+    private final NavigableMap<Long, Cancellable> alignmentTimers;
```
Same as above.

1996fanrui commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -335,6 +379,49 @@ public void checkpointState(
+
+            Cancellable currentAlignmentTimer =
+                    registerTimer.apply(
+                            () -> {
+                                try {
+                                    operatorChain.alignedBarrierTimeout(checkpointId);
+                                } catch (Exception e) {
+                                    ExceptionUtils.rethrowIOException(e);
+                                }
+                                synchronized (lock) {
+                                    alignmentTimers.remove(checkpointId);
+                                }
+                                return null;
+                            },
+                            Duration.ofMillis(timerDelay));
```
Same as why the `NavigableMap<Long, Cancellable> alignmentTimers` need to be guarded by lock?

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequest.java:
```diff
@@ -72,6 +74,51 @@ static ChannelStateWriteRequest write(
+                writer -> {
+                    try {
+                        List<Buffer> buffers = dataFuture.get();
+                        if (buffers == null || buffers.isEmpty()) {
```
Yes, I think there is no performance benefit for having null instead of `Collections.emptyList()` in this case (checkpoints are happening too rarely), while it's a bit safer (avoiding NPE)

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java:
```diff
@@ -240,9 +255,171 @@ private boolean processPriorityBuffer(BufferConsumer bufferConsumer, int partial
+                dataFuture);
+    }
+
+    private void checkChannelStateFutures(long currentCheckpointId) {
+        assert Thread.holdsLock(buffers);
+
+        while (!channelStateFutures.isEmpty()) {
+            Long checkpointId = channelStateFutures.firstKey();
+            if (checkpointId >= currentCheckpointId) {
+                break;
+            }
+            String exceptionMessage =
+                    String.format(
+                            "Received the barrier of checkpointId=%d, complete checkpointId=%d "
+                                    + "future by exception due to currently does not support "
+                                    + "concurrent unaligned checkpoints.",
+                            currentCheckpointId, checkpointId);
+            channelStateFutures
+                    .pollFirstEntry()
+                    .getValue()
+                    .completeExceptionally(new IllegalStateException(exceptionMessage));
+            LOG.info(exceptionMessage);
+        }
+    }
```
> I guess it might be supported in the future.

There are no plans for that. We can always reimplement it in the future, while for the time being we would have a simpler easier to read/maintain code. We just have to make sure that we put some `checkState()` or fail concurrent checkpoint here, that would ensure if someone implements/enables concurrent checkpoints, he won't accidentally miss place to modify.

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -103,6 +112,14 @@ class SubtaskCheckpointCoordinatorImpl implements SubtaskCheckpointCoordinator {
+    private final Clock clock;
+
+    /** Hold the AlignmentTimer for each checkpointId. */
+    @GuardedBy("lock")
```
What do you mean?

`SingleCheckpointBarrierHandler#resetAlignmentTimer` should only be called from the task thread as well.

However I don't see where `SubtaskCheckpointCoordinatorImpl#close()` is being called/used... That might be a bug.

1996fanrui replied:
Thanks for your reminder, I checked that the `SubtaskCheckpointCoordinatorImpl#close()` isn't called in anywhere. So can I delete the close method in this PR?

1996fanrui replied:
Hi @pnowojski , thanks for your reminder, I checked that the `SubtaskCheckpointCoordinatorImpl#close()` isn't called in anywhere. 

So could I fix the bug in other ticket? I can create a new JIRA. And I don't consider this multi-thread for alignmentTimer, right?

pnowojski replied:
> So could I fix the bug in other ticket? I can create a new JIRA.

Yes, that would be great. You could fix it in separate PR or as separate commit in this PR - whatever is more convenient for you :)

> And I don't consider this multi-thread for alignmentTimer, right?

alignmentTimer goes through the mailbox, as such it's also executed in the task thread, so this also doesn't require extra synchronisation.

1996fanrui replied:
I created FLINK-27724 just now, I will fix it after this PR.

https://issues.apache.org/jira/browse/FLINK-27724

1996fanrui commented:
> Thanks for the PR @1996fanrui ! I think the code looks good, I've left a couple of comments, mostly nits.

Hi @pnowojski , thanks a lot for your review, I will fix all comments asap.

1996fanrui commented:
Hi @pnowojski , I fixed your all comments and re-submit commits(refer the [coding style guideline](https://flink.apache.org/contributing/code-style-and-quality-pull-requests.html#separate-refactoring-cleanup-and-independent-changes)), please help to review again in your free time, thanks.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java:
```diff
@@ -112,6 +115,20 @@ public class PipelinedSubpartition extends ResultSubpartition
+
+    /**
+     * It is the checkpointId corresponding to channelStateFuture. And It should be always update
+     * with channelStateFuture.
```
nitty nit:
`{@link #channelStateFuture}`

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java:
```diff
@@ -240,9 +260,158 @@ private boolean processPriorityBuffer(BufferConsumer bufferConsumer, int partial
+                            String.format(
+                                    "%s has uncompleted channelStateFuture of checkpointId=%s, but it received "
+                                            + "a new timeoutable checkpoint barrier of checkpointId=%s, it maybe "
+                                            + "a bug due to currently does not support concurrent unaligned checkpoints.",
```
nit:
> a bug due to currently not supported concurrent unaligned checkpoint

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java:
```diff
@@ -240,9 +260,158 @@ private boolean processPriorityBuffer(BufferConsumer bufferConsumer, int partial
+                createChannelStateFuture(barrier.getId()));
+    }
+
+    private void completeTimeoutableCheckpointBarrier(BufferConsumer bufferConsumer) {
+        CheckpointBarrier barrier = parseAndCheckTimeoutableCheckpointBarrier(bufferConsumer);
+        if (channelStateFutureIsAvailable(barrier.getId())) {
+            completeChannelStateFuture(Collections.emptyList(), null);
+        }
+    }
```
Shouldn't we `checkState` here actually that `channelStateFutureIsAvailable(barrier.getId())` is true? Is there a valid scenario where this method should return false?

1996fanrui replied:
Actually, it may also be false here.

You mentioned in a [previous comment](https://github.com/apache/flink/pull/19723#discussion_r875800201). It happens when a checkpoint that has been aborted previously.

BTW, for these 2 comments, I'll add a brief description to the code explaining why we can't checkState, to prevent it from being corrected later.

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -335,6 +380,34 @@ public void checkpointState(
+            long checkpointId,
+            OperatorChain<?, ?> operatorChain,
+            CheckpointBarrier checkpointBarrier) {
+        if (alignmentTimer != null) {
```
shouldn't this be `checkState(alignmentTimer == null)`?

1996fanrui replied:
Actually, it may be null.

For a job without back pressure, the CP is completed quickly, but the timer is not triggered, so the next time the CP will continue, so this may be null. It should be reasonable to cancel the previous timer here.

pnowojski replied:
I see that in the end you have added the `checkState(alignmentTimer == null)` that I suggested?

1996fanrui replied:
Yes, I found some unit tests failed after added the `checkState(alignmentTimer == null)`. I have changed in my Mac, I will re-submit commits after everything  is ok.

pnowojski reviewed (commented):
Thanks for the update @1996fanrui . The code looks almost great. I've left a couple of more nits, and two questions about maybe adding a `checkState`.

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -234,6 +265,11 @@ public void abortCheckpointOnBarrier(
                 () -> operatorChain.broadcastEvent(new CancelCheckpointMarker(checkpointId)));
```
>  Should we move these 4 lines into runThrowing?
That's a good question. This method `abortCheckpointOnBarrier` is called only from the `CheckpointBarrierHandler`, which is only used with network input tasks, for which `actionExecutor` is always a no-op (it's being used only for legacy sources in `SourceStreamTask`, which doesn't have network input). So I think it doesn't matter really matter, but conceptually yes, it would be technically more correct to wrap all of those lines with `actionExecutor`. 

> And should we call cancelAlignmentTimer(); in close method?

Probably yes, to wake up anyone that might be waiting on the timer.

pnowojski reviewed (approved):
Really nicely done :) LGTM. And thank you for your patience and addressing the feedback.

edit: one more thing. Maybe we should also parametrise `UnalignedCheckpointStressITCase` to test for alignment timeout? `UnalignedCheckpointTestBase` and `org.apache.flink.streaming.util.TestStreamEnvironment#randomizeConfiguration` are adding quite a bit of test coverage for the timeout, but I remember `UnalignedCheckpointStressITCase` discovering quite a bit of bugs that the other tests haven't triggered/detected. What do you think?

1996fanrui commented:
> Really nicely done :) LGTM. And thank you for your patience and addressing the feedback.

Hi @pnowojski , It's my pleasure, and thanks a lot for your help.

Currently, I have addressed your all comments and 2 checkState. We may also need to discuss the issue of the thread-safe problem of alignmentTimer in close method.

Re `UnalignedCheckpointStressITCase`:
Currently, `TestStreamEnvironment#randomizeConfiguration` has unaligned timeout config. [code link](https://github.com/apache/flink/blob/3c5d3bacda580de1dc4986909a0aa5fc30e37885/flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/TestStreamEnvironment.java#L113), it's added in [FLINK-19682](https://issues.apache.org/jira/browse/FLINK-19682). And I think you are right, it's very useful during developing. It can find some easily overlooked boundary conditions.

pnowojski reviewed (commented):
> Re UnalignedCheckpointStressITCase:
> Currently, TestStreamEnvironment#randomizeConfiguration has unaligned timeout config. [code link](https://github.com/apache/flink/blob/3c5d3bacda580de1dc4986909a0aa5fc30e37885/flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/TestStreamEnvironment.java#L113), it's added in [FLINK-19682](https://issues.apache.org/jira/browse/FLINK-19682). And I think you are right, it's very useful during developing. It can find some easily overlooked boundary conditions.

I've thought that the timeout randomisation didn't work/was disabled in the `UnalignedCheckpointStressITCase`. But maybe indeed it's working there as it's not overwriting it. If that's the case, then all is fine.

1996fanrui commented:
> > Re UnalignedCheckpointStressITCase:
> > Currently, TestStreamEnvironment#randomizeConfiguration has unaligned timeout config. [code link](https://github.com/apache/flink/blob/3c5d3bacda580de1dc4986909a0aa5fc30e37885/flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/TestStreamEnvironment.java#L113), it's added in [FLINK-19682](https://issues.apache.org/jira/browse/FLINK-19682). And I think you are right, it's very useful during developing. It can find some easily overlooked boundary conditions.
> 
> I've thought that the timeout randomisation didn't work/was disabled in the `UnalignedCheckpointStressITCase`. But maybe indeed it's working there as it's not overwriting it. If that's the case, then all is fine.

It works due to set `checkpointing.randomization=true` in [pom.xml](https://github.com/apache/flink/blob/8cbb42a839919c3163f30928254a1afd0077f6a6/pom.xml#L1616).

1996fanrui commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -469,6 +550,7 @@ public void waitForPendingCheckpoints() throws Exception {
 
     @Override
     public void close() throws IOException {
+        cancelAlignmentTimer();
```
Hi @pnowojski , I'm agree with you. And I have a small question.

If the new `close()` method call `cancelAlignmentTimer()`, do we need to consider the thread-safe problem of alignmentTimer in close method. Although it doesn't matter much here.

pnowojski replied:
It shouldn't be necessary as long as you call `SubtaskCheckpointCoordinator#close()` from the task thread. `StreamTask#cleanUp` via registering to `resourceCloser` is probably the best place.

pnowojski commented:
All looks good, merging.
